// apollo-angular/packages/apollo-angular-link-http/build/src/HttpLink.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { ApolloLink, Observable } from 'apollo-link';
import { print } from 'graphql/language/printer';
import { normalizeUrl, mergeHeaders } from './utils';
// XXX find a better name for it
var HttpLinkHandler = /** @class */ (function (_super) {
    __extends(HttpLinkHandler, _super);
    function HttpLinkHandler(httpClient, options) {
        var _this = _super.call(this) || this;
        _this.options = options;
        _this.requester = new ApolloLink(function (operation) {
            return new Observable(function (observer) {
                var _a = operation.getContext(), headers = _a.headers, withCredentials = _a.withCredentials;
                var operationName = operation.operationName, variables = operation.variables, query = operation.query, extensions = operation.extensions;
                var body = {
                    operationName: operationName,
                    variables: variables,
                    query: print(query),
                };
                var postOptions = {
                    withCredentials: _this.options.withCredentials,
                    headers: _this.options.headers,
                };
                if (_this.options.includeExtensions) {
                    body.extensions = extensions;
                }
                if (typeof withCredentials !== 'undefined') {
                    postOptions.withCredentials = withCredentials;
                }
                // merge headers
                if (headers) {
                    postOptions.headers = mergeHeaders(postOptions.headers, headers);
                }
                var endpointURI = normalizeUrl(_this.options.uri);
                var defaultURI = 'graphql';
                var obs = httpClient.post(endpointURI || defaultURI, body, __assign({ observe: 'response', responseType: 'json', reportProgress: false }, postOptions));
                var sub = obs.subscribe({
                    next: function (result) {
                        observer.next(result.body);
                    },
                    error: function (err) {
                        observer.error(err);
                    },
                    complete: function () {
                        observer.complete();
                    },
                });
                return function () {
                    if (!sub.closed) {
                        sub.unsubscribe();
                    }
                };
            });
        }).request;
        return _this;
    }
    HttpLinkHandler.prototype.request = function (op) {
        return this.requester(op);
    };
    return HttpLinkHandler;
}(ApolloLink));
export { HttpLinkHandler };
var HttpLink = /** @class */ (function () {
    function HttpLink(httpClient) {
        this.httpClient = httpClient;
    }
    HttpLink.prototype.create = function (options) {
        return new HttpLinkHandler(this.httpClient, options);
    };
    HttpLink.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    HttpLink.ctorParameters = function () { return [
        { type: HttpClient, },
    ]; };
    return HttpLink;
}());
export { HttpLink };
